### PR TITLE
feat: refactor camp api

### PR DESCRIPTION
**Pull request by jjjding2**
## Notion ticket link

[Registration form - Camp API updates](https://www.notion.so/uwblueprintexecs/Tasks-38b49f83f6ce4b07ae188dabb37631e8?p=7ff572a8847548dd8e7f7491f4cb69aa)

## Implementation description
* Remove camperInfo and change to formQuestions
* Added in a new Validator for formQuestions
* Construct formQuestions, then baseCamps, then Camp
* Implemented exception guarantee

## Steps to test
1. {
"name": "jason-test",
"description": "'description'",
"location": "'knoxville'",
"capacity": 20,
"fee": 3000,
"formQuestions": [{
"type": "Text",
"question": "What is it",
"required": true,
"description": "Question question",
"options": ["one", "two"]
}],
"dates": ["12-2-2021", "12-30-2022"],
"startTime": "12:40",
"endTime": "16:50",
"ageLower": 12,
"ageUpper": 23,
"active": true
}

## What should reviewers focus on?
* createCamp implementation
* formQuestion and createCamp validators

## Checklist
- [x] My PR name is descriptive and in imperative tense
- [x] My commit messages are descriptive and in imperative tense. My commits are atomic and trivial commits are squashed or fixup'd into non-trivial commits
- [x] I have run the appropriate linter(s)
- [x] I have requested a review from the PL, as well as other devs who have background knowledge on this PR or who will be building on top of this PR

### CODE REVIEW

File: backend/typescript/services/implementations/campService.ts [hunk 1]
```diff
@@ -48,41 +38,76 @@ class CampService implements ICampService {
...
+        camp.formQuestions.map(async (formQuestion, i) => {
+          const question = await MgFormQuestion.create({
+            id: formQuestion.id,
+            type: formQuestion.type,
+            question: formQuestion.question,
+            required: formQuestion.required,
+            description: formQuestion.description,
+            options: formQuestion.options,
+          });
+          formQuestionIDs[i] = question._id;
```
Comment (annechung) [inline backend/typescript/services/implementations/campService.ts:56]: woah did not know you could just assign to whatever index in an array :o

File: backend/typescript/services/implementations/campService.ts [hunk 2]
```diff
@@ -48,41 +38,76 @@ class CampService implements ICampService {
...
+  async createCamp(camp: CreateCampDTO): Promise<CampDTO> {
+    let baseCamp: BaseCamp;
+    let newCamp: Camp;
+    const formQuestionIDs: string[] = [];
     try {
-      /* eslint no-underscore-dangle: 0 */
+      await Promise.all(
+        camp.formQuestions.map(async (formQuestion, i) => {
+          const question = await MgFormQuestion.create({
+            id: formQuestion.id,
```
Comment (annechung) [inline backend/typescript/services/implementations/campService.ts:49]: can we remove this line? id shouldn't be provided i don't think

**Review by annechung — COMMENTED** [review]
tested and looks good! few nits

**Review by annechung — APPROVED** [review]

File: backend/typescript/services/implementations/campService.ts (hunk 1 above)
Comment (jjjding2) [reply backend/typescript/services/implementations/campService.ts:56]: yeah LOL I had that same reaction

**Pull request merged**